### PR TITLE
build: Remove ceph-libboost* packages in install-deps

### PR DESCRIPTION
**Pull request by adamemerson**
https://tracker.ceph.com/issues/62097

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by adamemerson** [comment]
The third commit may not be worth it, as there are also incompatibilities between Asio as used in Quincy and newer Boost.Asio versions.

**Comment by ljflores** [comment]
jenkins test api

**Review by idryomov — COMMENTED** [review]
Why is `clean_boost_on_ubuntu` being added directly to quincy?  Since Boost will inevitably get bumped again in the future, wouldn't it be useful on all branches?  Then this change could be backported following the usual workflow too.

The third commit indeed seems unnecessary.  With quincy staying on 1.75, adding checks for 1.81 or 1.82 only contributes to the confusion.

**Comment by idryomov** [comment]
> > Why is `clean_boost_on_ubuntu` being added directly to quincy? Since Boost will inevitably get bumped again in the future, wouldn't it be useful on all branches? Then this change could be backported following the usual workflow too.
> 
> Because I was trying to hammer out a working quincy build and I thought it would be faster to get it fixed that way. If people would rather it be against main, I'm happy to do that instead.

> Because I was trying to hammer out a working quincy build and I thought it would be faster to get it fixed that way. If people would rather it be against main, I'm happy to do that instead.

This is breaking the workflow, only marginally faster and rarely pays out in the long run.  I'd vote for adding the cleanup step in main and backporting to pacific, quincy and reef.

> 
> However, my preferred solution would be to have Jenkins do more cleanup between builds so the state of the builder isn't quite so chaotic, and if that were done we wouldn't really need it. So I was hoping to hear something on that matter from the infrastructure team and whether it was something we could aim for.

https://lists.ceph.io/hyperkitty/list/dev@ceph.io/message/OAH633PLMDWQIO5RTC5NAGVEABVCUKDV/

File: install-deps.sh [hunk 1]
```diff
@@ -146,9 +146,29 @@ function install_pkg_on_ubuntu {
     fi
 }
 
+boost_ver=1.82
+
+function clean_boost_on_ubuntu {
```
Comment (idryomov) [inline install-deps.sh:151]: It looks like `install_boost_on_ubuntu()` already has some code for removing undesired packages.  Can you summarize why it's not working in this case?

Did you consider expanding it instead of duplicating some of it in a new function?

Comment (adamemerson) [reply install-deps.sh:151]: Two reasons. First, it wasn't being called because in Quincy we didn't have a case for Jammy. Since under the proposed Quincy fix Jammy doesn't call `install_boost_on_ubuntu()` (as it does not use the ceph-libboost packages), we need a separate clean function.

Second, the file it removes from `/etc/apt/sources.list.d` is not actually the file `install_boost_on_ubuntu()` installs there.

I just repushed to remove the relevant stuff from `install_boost_on_ubuntu()` though, so it's not duplicated.

Comment (tchaikov) [reply install-deps.sh:151]: the commit message reads

> To force removal of overly new libboost packages when building with system boost.

but i see there are couple things in this change


1. extract "clean_boost_on_ubuntu()" out of "install_boost_on_ubuntu()" // you explained this in the comment above. but it's not explained in the commit message.
2. always call "install_pkg_on_ubuntu()" in "install_boost_on_ubuntu()" even if some boost package(s) is found with the expected version
3. always call "clean_boost_on_ubuntu()"
4. remove more source lists in `clean_boost_on_ubuntu()` // i can guess, but i don't really understand
5. list the source list  // for debugging?
6. call `apt-get update` in `clean_boost_on_ubuntu()` // what is it for?
7. remove all liboost `xargs` // i guess it's an improvement, but it deserves some explanation as well

in short, the change is dense. and i failed to read between the lines. could you please put more details on why and why and why in the commit message?

Comment (adamemerson) [reply install-deps.sh:151]: It got more complicated pulling it onto main. Really I think it should have just been done against Quincy and that's it because nothing after Quincy ever had the problem of there NOT being a call to `install_boost_on_ubuntu()` for Jammy

Comment (idryomov) [reply install-deps.sh:151]: > Really I think it should have just been done against Quincy and that's it because nothing after Quincy ever had the problem of there NOT being a call to `install_boost_on_ubuntu()` for Jammy

`make check` on pacific is affected the same way.  I don't know if the root cause is the same, but it's not a quincy-only issue.

Also, is there anything to guarantee that this problem would not repeat with a different combination of Ceph and Ubuntu LTS releases in the future?

Comment (adamemerson) [reply install-deps.sh:151]: I believe a straightforward backport of both commits needed for Quincy would fix Pacific as well.

Putting this on main wouldn't stop problems in stable releases if we switch LTS's either. It would mean they have the wrong boost removed, but nothing installs the right boost, because there's no case for the new LTS in the old install-deps.sh, and no version of an old boost package built against it.

At present it looks like if we *just* remove the wrong boost from Quincy and push then jenkins will fetch the boost source and build it, which I honestly didn't expect to work and I wouldn't really count on working in the future.

Comment (adamemerson) [reply install-deps.sh:151]: >     2. always call "install_pkg_on_ubuntu()" in "install_boost_on_ubuntu()" even if some boost package(s) is found with the expected version

This was a mistake, thank you.

>     5. list the source list  // for debugging?

For debugging, yes, but I didn't mean to leave it in. Thank you.

Comment (idryomov) [reply install-deps.sh:155]: > At present it looks like if we _just_ remove the wrong boost from Quincy and push then jenkins will fetch the boost source and build it, which I honestly didn't expect to work and I wouldn't really count on working in the future.

I think this is expected and actually _can_ be counted upon.  At a high level:
- we try system boost
- if system boost isn't good enough, we try "ceph" pre-built boost
- if "ceph" pre-built boost isn't there/good enough, we download and build from scratch

When the build that you produced for 1.82 got pruned (I think that is what happened?), `make check` continued working precisely because of that.

Comment (adamemerson) [reply install-deps.sh:155]: In general, yes.

But since more and more people are locking down their builders' ability to access the Internet, I don't expect it to keep working indefinitely.

But I don't care, I'll do either one. I just want to not leave quincy and pacific builders broken longer than necessary.

File: install-deps.sh [hunk 2]
```diff
@@ -146,21 +146,34 @@ function install_pkg_on_ubuntu {
...
 function install_boost_on_ubuntu {
     ci_debug "Running install_boost_on_ubuntu() in install-deps.sh"
-    local ver=1.82
-    local installed_ver=$(apt -qq list --installed ceph-libboost*-dev 2>/dev/null |
-                              grep -e 'libboost[0-9].[0-9]\+-dev' |
-                              cut -d' ' -f2 |
-                              cut -d'.' -f1,2)
+    local installed_ver=$boost_installed_ver
     if test -n "$installed_ver"; then
         if echo "$installed_ver" | grep -q "^$ver"; then
```
Comment (idryomov) [inline install-deps.sh:172]: This check seems broken as `$ver` variable definition has been moved.

File: install-deps.sh [hunk 3]
```diff
@@ -146,21 +146,34 @@ function install_pkg_on_ubuntu {
...
 }
 
+boost_ver=1.82
+boost_installed_ver=$(apt -qq list --installed 'ceph-libboost*-dev' 2> /dev/null \
+			  | grep -e 'libboost[0-9].[0-9]\+-dev' \
+			  | cut -d ' ' -f 2 \
+			  | cut -d '.' -f1,2)
+
+function clean_boost_on_ubuntu {
+    local ver=$boost_ver
```
Comment (idryomov) [inline install-deps.sh:156]: ... and here `$ver` variable isn't used.

File: install-deps.sh [hunk 4]
```diff
@@ -146,21 +146,34 @@ function install_pkg_on_ubuntu {
     fi
 }
 
+boost_ver=1.82
+boost_installed_ver=$(apt -qq list --installed 'ceph-libboost*-dev' 2> /dev/null \
```
Comment (idryomov) [inline install-deps.sh:150]: Wouldn't this attempt to run `apt` on e.g. CentOS?

File: install-deps.sh [hunk 5]
```diff
@@ -146,21 +146,34 @@ function install_pkg_on_ubuntu {
...
+			  | grep -e 'libboost[0-9].[0-9]\+-dev' \
+			  | cut -d ' ' -f 2 \
+			  | cut -d '.' -f1,2)
+
+function clean_boost_on_ubuntu {
+    local ver=$boost_ver
+    local installed_ver=$boost_installed_ver
+    $SUDO rm -f "/etc/apt/sources.list.d/ceph-libboost${installed_ver}.list"
+    $SUDO rm -f "/etc/apt/sources.list.d/ceph-libboost.list"
+    $SUDO rm -f "/etc/apt/sources.list.d/libboost.list"
```
Comment (idryomov) [inline install-deps.sh:160]: > As the sources.list.d being removed in the original cleanup code isn't the one we're currently installing in the install code, add a removal for the currently used source [...]

Perhaps forestalling another set of questions from Kefu, I'm confused by this explanation.  The original cleanup code removes the version that is being installed, not the one that is intended to be installed -- it removes `ceph-libboost.*${installed_ver}.*` which is very similar to the new code removing `ceph-libboost${installed_ver}.list`.

- Under what conditions would `ceph-libboost.*${installed_ver}.*` not match?  What would the mismatch be exactly?
- What installs `ceph-libboost.list` file?
- What installs `libboost.list` file (it doesn't even have `ceph-` prefix)?

Comment (adamemerson) [reply install-deps.sh:160]: > > As the sources.list.d being removed in the original cleanup code isn't the one we're currently installing in the install code, add a removal for the currently used source [...]
> 
> Perhaps forestalling another set of questions from Kefu, I'm confused by this explanation. The original cleanup code removes the file for the version that is already installed -- it removes `ceph-libboost.*${installed_ver}.*` which is very similar to the new code removing `ceph-libboost${installed_ver}.list`.
> 
>     * Under what conditions would `ceph-libboost.*${installed_ver}.*` not match?  What would the mismatch be exactly?

I don't know. I'm not certain that we couldn't have two ceph-libboost.*-dev packages installed and since the idea was to clean things up so it wouldn't pull in the wrong thing, I didn't think it was safe to assume there was exactly one installed version.
 
>     * What installs `ceph-libboost.list` file?

Nothing now. I removed that removal.
 
>     * What installs `libboost.list` file (it doesn't even have `ceph-` prefix)?

Line 143 in `install_pkg_on_ubuntu()` with project set to `libboost` as it is in `install_boost_on_ubuntu()`.

File: install-deps.sh (hunk 2 above)
Comment (adamemerson) [reply install-deps.sh:174]: Fixed

File: install-deps.sh [hunk 6]
```diff
@@ -146,19 +146,32 @@ function install_pkg_on_ubuntu {
     fi
 }
 
+boost_ver=1.82
+
+function clean_boost_on_ubuntu {
```
[the remark this answers is not in the document]
Comment (adamemerson) [reply install-deps.sh:151]: Done.

File: install-deps.sh [hunk 7]
```diff
@@ -146,19 +146,32 @@ function install_pkg_on_ubuntu {
...
+    local installed_ver=$(apt -qq list --installed ceph-libboost*-dev 2>/dev/null |
+                              grep -e 'libboost[0-9].[0-9]\+-dev' |
+                              cut -d' ' -f2 |
+                              cut -d'.' -f1,2)
+    $SUDO rm -f "/etc/apt/sources.list.d/ceph-libboost${installed_ver}.list"
```
[the remark this answers is not in the document]
Comment (adamemerson) [reply install-deps.sh:156]: Does nothing, it prints stuff then exits successfully.

[the remark this answers is not in the document]
Comment (adamemerson) [reply install-deps.sh]: In case how we query `installed_ver` breaks I'm using it *only* to decide whether to bail out early in `clean_boost_on_ubuntu()` and then only if it has no spaces. It still functions as it did in `install_boost_on_ubuntu()` because by that point I actually am confident there is zero or one version installed.

File: install-deps.sh [hunk 8]
```diff
@@ -146,19 +146,51 @@ function install_pkg_on_ubuntu {
...
+    # Historical packages
+    $SUDO rm -f "/etc/apt/sources.list.d/ceph-libboost*.list"
+    # Currently used
+    $SUDO rm -f "/etc/apt/sources.list.d/libboost.list"
+    # Refresh package list so things aren't in the available list.
+    $SUDO apt-get update -y
+    # Remove all ceph-libboost dev packages except, possibly, those of
+    # the version we currently want.
+    apt -qq list --installed 'ceph-libboost*-dev' 2> /dev/null | \
+	grep -F -v "$boost_ver" | \
```
Comment (idryomov) [inline install-deps.sh:177]: Now that a check for the installed version has been added at the top, do we still need this filter here?  Control gets to this point if either a) no boost packages are installed, b) a single version but not the one desired is installed or c) multiple versions are installed.  Since we are removing all related repositories, it feels weird to potentially leave some packages installed from those repositories behind.

I would treat (c) as a "something got messed up around here" case and remove all related packages, letting `install_boost_on_ubuntu()` install the desired version from scratch.

File: install-deps.sh [hunk 9]
```diff
@@ -146,19 +146,51 @@ function install_pkg_on_ubuntu {
...
+    # multiple versions installed.
+    local installed_ver=$(apt -qq list --installed ceph-libboost*-dev 2>/dev/null |
+                              grep -e 'libboost[0-9].[0-9]\+-dev' |
+                              cut -d' ' -f2 |
+                              cut -d'.' -f1,2)
+    # If installed_ver contains a space, we can't really count on it,
+    # but otherwise, bail out if the version installed is the version
+    # we want.
+    if test -n "$installed_ver" && echo "$installed_ver" | grep -v -q ' '; then
+        if echo "$installed_ver" | grep -q "^$ver"; then
```
[the remark this answers is not in the document]
Comment (adamemerson) [reply install-deps.sh:163]: Done.

File: install-deps.sh [hunk 10]
```diff
@@ -146,19 +146,49 @@ function install_pkg_on_ubuntu {
...
+    fi
+
+    # Historical packages
+    $SUDO rm -f "/etc/apt/sources.list.d/ceph-libboost*.list"
+    # Currently used
+    $SUDO rm -f "/etc/apt/sources.list.d/libboost.list"
+    # Refresh package list so things aren't in the available list.
+    $SUDO apt-get update -y
+    # Remove all ceph-libboost dev packages except, possibly, those of
+    # the version we currently want.
```
Comment (idryomov) [inline install-deps.sh:175]: This comment became outdated with the removal of the filter.

File: install-deps.sh [hunk 11]
```diff
@@ -146,19 +146,49 @@ function install_pkg_on_ubuntu {
...
+    apt -qq list --installed 'ceph-libboost*-dev' 2> /dev/null | \
+	cut -f 1 -d / | \
+	xargs $SUDO apt-get -y remove
```
Comment (idryomov) [inline install-deps.sh:178]: With the filter removed, do we still need `cut` and `xargs` or would the current `$SUDO env DEBIAN_FRONTEND=noninteractive apt-get -y remove "ceph-libboost.*${installed_ver}.*"` (note wildcards) do the job?

Comment (adamemerson) [reply install-deps.sh:178]: We don't want `${installed_ver}`. But I think `ceph-libboost.*` should be fine.

I had originally intended to keep it as `-dev` as the minimal conservative operation that would unbreak the build, but on consideration I don't think there's any reason to spare the non-dev packages after we're removing the repos they came from, and having them doesn't buy us anything.

File: install-deps.sh [hunk 12]
```diff
@@ -146,19 +146,48 @@ function install_pkg_on_ubuntu {
...
+    # Find currently installed version. This may not work if there are
+    # multiple versions installed.
+    local installed_ver=$(apt -qq list --installed ceph-libboost*-dev 2>/dev/null |
+                              grep -e 'libboost[0-9].[0-9]\+-dev' |
+                              cut -d' ' -f2 |
+                              cut -d'.' -f1,2)
+    # If installed_ver contains a space, we can't really count on it,
+    # but otherwise, bail out if the version installed is the version
+    # we want.
+    if test -n "$installed_ver" && echo "$installed_ver" | grep -v -q ' '; then
```
Comment (idryomov) [inline install-deps.sh:162]: Are you sure that `installed_ver` would actually contain spaces?  I just tried a similar command and it's newlines instead of spaces.  This matters since the variable expansion is quoted here:
```
$ foo=$(apt -qq list --installed emacs* 2>/dev/null | cut -d' ' -f2 | cut -d'.' -f1,2)
$ echo $foo
1:26.3+1-1ubuntu2 1:26.3+1-1ubuntu2 1:26.3+1-1ubuntu2 1:26.3+1-1ubuntu2 3.0
$ echo "$foo"
1:26.3+1-1ubuntu2
1:26.3+1-1ubuntu2
1:26.3+1-1ubuntu2
1:26.3+1-1ubuntu2
3.0
```

File: install-deps.sh [hunk 13]
```diff
@@ -146,19 +146,48 @@ function install_pkg_on_ubuntu {
...
+    # Historical packages
+    $SUDO rm -f "/etc/apt/sources.list.d/ceph-libboost*.list"
+    # Currently used
+    $SUDO rm -f "/etc/apt/sources.list.d/libboost.list"
+    # Refresh package list so things aren't in the available list.
+    $SUDO apt-get update -y
+    # Remove all ceph-libboost packages. We have an early return if
+    # the desired version is already (and the only) version installed,
+    # so no need to spare it.
+    $SUDO env DEBIAN_FRONTEND=noninteractive apt-get -y remove "ceph-libboost.*"
```
Comment (idryomov) [inline install-deps.sh:177]: Nit: do `apt` and `apt-get` require different patterns?  Just noticing `.*` here vs `*` above when assigning `installed_ver` supposedly expressing the same thing.

File: install-deps.sh (hunk 12 above)
Comment (idryomov) [reply install-deps.sh:162]: And further, since there would be multiple packages with the same version (atomic, chrono, container, etc), wouldn't this condition always trigger (assuming the separator that it's looking for is correct)?

File: install-deps.sh (hunk 13 above)
Comment (adamemerson) [reply install-deps.sh:177]: No. it tries globs first then regular expressions. But no reason not to switch it to a glob.

I'd been preferring apt-get whenever possible since apt has a warning against using it in scripts.

File: install-deps.sh (hunk 12 above)
Comment (adamemerson) [reply install-deps.sh:162]: No. The grep filters them out.

Comment (adamemerson) [reply install-deps.sh:162]: You are correct about the space. Switching to `[[:space:]]`

File: install-deps.sh (hunk 13 above)
Comment (adamemerson) [reply install-deps.sh:177]: But a glob or regex it can't match will error out in a way that removing an empty package list won't, so this needs a guard.

File: install-deps.sh (hunk 12 above)
Comment (adamemerson) [reply install-deps.sh]: The grep does not actually filter what we'd like, though, so I'll use uniq.

File: install-deps.sh [hunk 14]
```diff
@@ -146,19 +146,49 @@ function install_pkg_on_ubuntu {
...
+    ci_debug "Running clean_boost_on_ubuntu() in install-deps.sh"
+    # Find currently installed version. This may not work if there are
+    # multiple versions installed.
+    local installed_ver=$(apt -qq list --installed ceph-libboost*-dev 2>/dev/null |
+                              cut -d' ' -f2 |
+                              cut -d'.' -f1,2 | sort | uniq)
+    # If installed_ver contains whitespace, we can't really count on it,
+    # but otherwise, bail out if the version installed is the version
+    # we want.
+    if test -n "$installed_ver" && echo "$installed_ver" | grep -v -q '[[:space:]]'; then
```
Comment (idryomov) [inline install-deps.sh:161]: I don't think this works as expected.  First, `echo` appends a newline, so there is always going to be a newline there even with a single version.  Second, `grep` is line-based, so it wouldn't match newlines unless special care is taken (like -z/-Z options of GNU grep).

Please test your changes at least on some toy examples before pushing -- I have lost count of the number of silly bugs like this found in this function in review.  For a proper test, I would suggest setting up some Ubuntu containers, adding those shaman repos and running `install-deps.sh` through scenarios:
- no boost packages installed
- one version installed, matches the desired version
- one version installed, doesn't match the desired version
- two versions installed, one matches the desired version
- two versions installed, none match the desired version

File: install-deps.sh [hunk 15]
```diff
@@ -146,19 +146,49 @@ function install_pkg_on_ubuntu {
...
+        if echo "$installed_ver" | grep -q "^$boost_ver"; then
+            return
+        fi
+    fi
+
+    # Historical packages
+    $SUDO rm -f "/etc/apt/sources.list.d/ceph-libboost*.list"
```
[the remark this answers is not in the document]
Comment (adamemerson) [reply install-deps.sh:170]: Fixed. Misread this comment the first time as applying to the apt-get remove.

File: install-deps.sh [hunk 16]
```diff
@@ -146,19 +146,51 @@ function install_pkg_on_ubuntu {
...
+	    return
+        fi
+    fi
+
+    # Historical packages
+    $SUDO rm -f "/etc/apt/sources.list.d/ceph-libboost*.list"
+    # Currently used
+    $SUDO rm -f "/etc/apt/sources.list.d/libboost.list"
+    # Refresh package list so things aren't in the available list.
+    $SUDO apt-get update -y
```
[the remark this answers is not in the document]
Comment (adamemerson) [reply install-deps.sh:174]: No principled reason, I'll add it.

**Comment by adamemerson** [comment]
jenkins, retest this please

**Comment by cbodley** [comment]
jenkins test make check

**Comment by cbodley** [comment]
sorry i haven't kept up with review, but this looks very reasonable to me in its current form 👍

File: install-deps.sh [hunk 17]
```diff
@@ -146,19 +146,51 @@ function install_pkg_on_ubuntu {
...
+	    return
+        fi
+    fi
+
+    # Historical packages
+    $SUDO rm -f /etc/apt/sources.list.d/ceph-libboost*.list
+    # Currently used
+    $SUDO rm -f /etc/apt/sources.list.d/libboost.list
+    # Refresh package list so things aren't in the available list.
+    $SUDO env DEBIAN_FRONTEND=noninteractive apt-get update -y || true
```
Comment (idryomov) [inline install-deps.sh:174]: > Copying the way we do apt-get update elsewhere.

Wouldn't it have been `$SUDO env DEBIAN_FRONTEND=noninteractive apt-get update -y -o Acquire::Languages=none -o Acquire::Translation=none || true` then, the way it's done elsewhere in `install-deps.sh` and also in `qa/workunits/ceph-helpers-root.sh`?  These `-o` options are supposed to speed things up a bit, see 45781602c9a02bd30986ea59a5603deaebf9c435.

**Review by idryomov — COMMENTED** [review]
Otherwise LGTM!

File: install-deps.sh (hunk 17 above)
Comment (idryomov) [reply install-deps.sh:174]: Ah, I see that `install-deps.sh` has an instance without these options too, so it's consistent already (didn't scroll far enough...).  Feel free to disregard.

**Comment by idryomov** [comment]
jenkins test make check

**Comment by idryomov** [comment]
jenkins test windows

**Comment by idryomov** [comment]
jenkins test windows

**Review by idryomov — APPROVED** [review]
